### DISCUSSION TITLE
Word/symbol expansion

### DISCUSSION THREAD

**Discussion by mclearc** [meow-edit/meow#173] [Q&A]
This may be due to my ignorance of the finer details of meow, but I can't seem to do the following:

take a word/symbol pair like `@kitcher2018`, which is pretty standard for of citation that I use. If I place the point on `@` and call `meow-next-word` I'll get `kitcher2018` highlighted. Now let's suppose that I *also* want the `@` highlighted. How can I easily do that? Calling either `meow-mark-word` or `meow-mark-symbol` won't work. And `meow-block` doesn't do what I would expect here either, since it just jumps the point to a bracketed expression rather than expanding the block to the smallest next larger increment (in this case `@kitcher2018`). And `meow-bounds/inner-of-thing` doesn't help either. Any suggestions on what to do here? Could `meow-block` be modified?

**Comment by velppa** [comment]
For such cases I use `meow-mark-word` `meow-reverse` `meow-left-expand`, which are `w;H`
for me. Not sure whether it's the most effective approach.

Related, given the text with `|` denoting cursor:

    (format |"%s %d" font size)

`meow-next-word` would select `%s`, but if it's a single-quote instead of
double-quote, it would select it as well `'%s`. Is it expected?

    (format |'%s %d' font size)

**Reply by DogLooksGood** [reply]
Emmm, in elisp, strings can't be quoted by single quote.
But it should still select `%s`, because word selection only include characters those recognized as `-_w` in syntax table(whitespaces/symbol constituent/word constituent).
A simple `skip-syntax-forward/backward` [is used here](https://github.com/meow-edit/meow/blob/master/meow-command.el#L726).

In your case, `'%s` is selected, probably because `'` is added as one of `-_w` in syntax table.  By default, it belongs to `prefix`, you can check the syntax table with `C-h s`.

**Reply by velppa** [reply]
Thanks for sharing, didn't know about syntax tables. Surely, in Elisp strings aren't single quoted, the example was purely illustrative as I bumped into it in some other mode.

**Comment by eshrh** [comment]
Sorry, maybe I don't understand...

If you have (| is the cursor) `@|kitcher2018` and call meow-next word (e), you get `kitcher2018|` highlighted (identical behavior to meow-mark-word here). Then calling `meow-mark-symbol` (W) seems to make `@kitcher2018|` highlighted on my computer. Is there something wrong about this?

**Reply by mclearc** [reply]
calling `meow-mark-symbol` doesn't highlight `@kitcher2018` but only `kitcher2018`. Perhaps there has been a change in the last day or so that would explain this? Otherwise I'm not sure....

**Reply by mclearc** [reply]
Yeah I'm on the most recent commit and the process you describe doesn't work for me.

**Reply by eshrh** [reply]
Sorry, this is my fault -- I was on an older commit (forgot to rebuild) and it used to work... and now I've lost which commit I was on :smiling_face_with_tear:.

**Reply by eshrh** [reply]
Ok nevermind, I wasn't hallucinating. The problem is in the built-in function `bounds-of-thing-at-point`, the bounds it returns for 'word and 'symbol differ based on mode. I tested your example in a scratch buffer which used lisp interaction mode, and the @ symbol is included in the 'symbol thing. In the fundamental and org mode buffers it's not

That's actually really annoying, I'll look into it.

**Reply by mclearc** [reply]
great - thanks. I was testing this in fundamental, org,  and markdown mode. Hadn't thought that it might vary all that much.

**Reply by mclearc** [reply]
Given Meow's (laudable) tendency to defer to underlying emacs operations II assume that this is going to amount to a difference between programming modes and text modes like org and markdown. But it would be great to strive for as much continuity as possible across modes. Principle of least surprise is great here. So even if nothing is to be done about it maybe saying something in the README about variances would be helpful.

**Reply by eshrh** [reply]
I mean, if you want a quick fix, you can execute something like this:

`(modify-syntax-entry ?@ "_" org-mode-syntax-table)`

This sets the character @ to be recognized as a symbol which is denoted by _ in the syntax table. You can do this for the markdown syntax table as well. You might consider putting a function that does this into something like `after-change-major-mode-hook` for global behavior. 

I don't think this is really nice though, but I'm not sure if there's a good solution. I'm considering making all meow commands use their own syntax table where we take care of these problems, but I'm don't know whether this could lead to other problems. A really basic idea would be to wrap every call to `bounds-of-thing-at-point` with `with-syntax-table`.

I totally agree with you that this is pretty surprising, and I think something should be done about it -- so I'm open to ideas.

**Reply by DogLooksGood** [reply]
@mclearc You are right, there are some details not mentioned by the document.  Syntax table controls what characters are included in word selection.  It is also the way builtin `forward-word`/`backward-word` works. Meow respects the builtins whenever it's possible.

In your case, modifying syntax table is the best solution.  Syntax table is the perfect solution for people who want to tweak the underlying behavior.  You will like it once you get used to it.

**Reply by mclearc** [reply]
I think it is totally reasonable if you want the user to make their own modifications here, especially since meow doesn't intend to make any changes to the underlying emacs functioning. But it might be useful to flag this clearly in the README info, since this won't be at all obvious to people switching from other modal bindings, especially from evil.

**Reply by DogLooksGood** [reply]
Yes, I agree.  But it's not easy.  When I switched from evil to vanilla, I felt like I have to relearn everything.  Since evil made a context which do not need its users to know much about Emacs itself.  And Emacs itself is complex too, by said Emacs is customizable, as long as you have new perceptions, you can have new customizations.  It will be too overcrowding to list all potentially useful information in Meow's documentation, It's also hard for me to figure out what is worth to be listing.  If we want a good documentation for evil users, that must be a migration guide, from the perspective of evil.  Of course, it will take a lot of time.  At the moment, I am not ready for that, there is still room for other improvements. Currently, I'm focusing on making Meow a good modal editing package at least for people who is familiar with vanilla Emacs.

**Reply by eshrh** [reply]
@DogLooksGood 

I think it's still valuable to offer an option to use the same syntax table across modes. Here's what I'm thinking:

Offer customization variable, `meow-syntax-table`, initially set to nil. Every call to bounds-of-thing-at-point and other thingatpt.el functions is wrapped with `(with-syntax-table)`. If the variable is nil, then it uses the syntax table of the current mode, default behavior. Otherwise, it uses the given syntax table.

**Reply by DogLooksGood** [reply]
@eshrh It doesn't feel right to me to introduce another variable.  `modify-syntax-entry` can either set globally or set for a specific syntax table of a major-mode. I think it provides everything needed for a same syntax table cross modes. And of course it works with all commands. Our `meow-syntax-table` can only be used in meow, will not work with builtin commands. 

Why there's no such single global syntax table? Just because it's not a good idea. 
For examples,  
- `a-b-c` in Lisp / C
- `~a` in Elisp / Clojure
- `'a'` in Lisp / Python

There are just so many differences.
